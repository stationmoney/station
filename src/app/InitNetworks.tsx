import { PropsWithChildren, useEffect, useState } from "react"
import axios from "axios"
import { ASSETS } from "config/constants"
import createContext from "utils/createContext"
import NetworkLoading from "./NetworkLoading"

export const [useNetworks, NetworksProvider] = createContext<{
  networks: InterchainNetworks
  filterEnabledNetworks: <T>(network: Record<string, T>) => Record<string, T>
}>("useNetworks")

const InitNetworks = ({ children }: PropsWithChildren<{}>) => {
  const [networks, setNetworks] = useState<InterchainNetworks>()
  const [enabledNetworks, setEnabledNetworks] = useState<string[]>([])

  useEffect(() => {
    const fetchChains = async () => {
      const { data: chains } = await axios.get<InterchainNetworks>(
        "/station/chains.json",
        {
          baseURL: ASSETS,
        }
      )
      setNetworks(chains)
    }

    fetchChains()
  }, [])

  useEffect(() => {
    const testChains = async () => {
      if (!networks) return
      const testBase = {
        ...networks.mainnet,
        ...networks.testnet,
        ...networks.classic,
      }

      const result = await Promise.all(
        Object.values(testBase).map(async (network) => {
          try {
            const { data } = await axios.get(
              "cosmos/base/tendermint/v1beta1/node_info",
              {
                baseURL: network.lcd,
                timeout: 5_000,
              }
            )
            return data.default_node_info.network as string
          } catch (e) {
            return null
          }
        })
      )

      setEnabledNetworks(
        result.filter((r) => typeof r === "string") as string[]
      )
    }

    testChains()
  }, [networks])

  if (!networks || !enabledNetworks.length)
    return <NetworkLoading title="Connecting to available networks..." />

  return (
    <NetworksProvider
      value={{
        networks,
        filterEnabledNetworks: (networks) =>
          Object.fromEntries(
            Object.entries(networks).filter(([chainID]) =>
              enabledNetworks.includes(chainID)
            )
          ),
      }}
    >
      {children}
    </NetworksProvider>
  )
}

export default InitNetworks
