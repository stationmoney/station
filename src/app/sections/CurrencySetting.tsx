import { useSupportedFiat } from "data/queries/coingecko"
import { useCurrencyState } from "data/settings/Currency"
import { RadioGroup } from "components/form"
import WithSearchInput from "pages/custom/WithSearchInput"

const CurrencySetting = () => {
  const { data: fiatList = [] } = useSupportedFiat()
  const [currency, setCurrency] = useCurrencyState()

  return (
    <WithSearchInput gap={8} small>
      {(input) => (
        <RadioGroup
          options={fiatList
            .filter(
              (fiat) =>
                fiat.name.toLowerCase().includes(input.toLowerCase()) ||
                fiat.id.toLowerCase().includes(input.toLowerCase()) ||
                fiat.symbol.toLowerCase().includes(input.toLowerCase())
            )
            .map((fiat) => {
              return { value: fiat.id, label: `${fiat.symbol} - ${fiat.name}` }
            })}
          value={currency.id}
          onChange={(value) => {
            setCurrency(
              fiatList.find((fiat) => fiat.id === value) as {
                name: string
                symbol: string
                id: string
              }
            )
          }}
        />
      )}
    </WithSearchInput>
  )
}

export default CurrencySetting
