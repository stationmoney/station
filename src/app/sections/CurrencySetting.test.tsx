import React from "react"
import { render, screen } from "@testing-library/react"
import CurrencySetting from "./CurrencySetting"
import { RecoilRoot } from "recoil"
import { NetworksProvider } from "../InitNetworks"
import renderer from "react-test-renderer"
import user from "@testing-library/user-event"

type TokenFilter = <T>(network: Record<string, T>) => Record<string, T>
const networks = {} as jest.Mocked<InterchainNetworks>
const mockedTokenFilter = {} as jest.Mocked<TokenFilter>

function renderComponent() {
  return render(
    <NetworksProvider
      value={{
        networks: networks,
        filterEnabledNetworks: mockedTokenFilter,
        filterDisabledNetworks: mockedTokenFilter,
      }}
    >
      <RecoilRoot>
        <CurrencySetting />
      </RecoilRoot>
    </NetworksProvider>
  )
}

jest.mock("../../data/queries/coingecko", () => {
  // Retain functionality of other exports.
  const actual = jest.requireActual("../../data/queries/coingecko")

  // Replace functionality of useSupportedFiat.
  const useSupportedFiatMock = () => {
    return {
      data: supportedFiatListMock,
    }
  }

  return {
    ...actual,
    useSupportedFiat: useSupportedFiatMock,
  }
})

// Mocked supported fiats list.
const supportedFiatListMock = [
  {
    id: "JPY",
    name: "Japanese Yen",
    symbol: "¥",
  },
  {
    id: "NGN",
    name: "Nigerian Naira",
    symbol: "₦",
  },
  {
    id: "USD",
    name: "United States Dollar",
    symbol: "$",
  },
]

describe("Currency Setting loads", () => {
  it("only currencies in supported fiat list display", () => {
    renderComponent()

    // Ensure all currencies in supported fiat list are displayed.
    for (let { name, symbol } of supportedFiatListMock) {
      let currencyDisplay = screen.getByText(`${symbol} - ${name}`)
      expect(currencyDisplay).toBeInTheDocument()
    }

    // Ensure only supported fiat list currencies are on display.
    const buttons = screen.getAllByRole("button")
    expect(buttons).toHaveLength(3)
  })

  // Ensure snapshot matches original component.
  it("matches snapshot", () => {
    const tree = renderer
      .create(
        <NetworksProvider
          value={{
            networks: networks,
            filterEnabledNetworks: mockedTokenFilter,
            filterDisabledNetworks: mockedTokenFilter,
          }}
        >
          <RecoilRoot>
            <CurrencySetting />
          </RecoilRoot>
        </NetworksProvider>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe("User can change currencies", () => {
  it("allows user to change currencies", async () => {
    const { asFragment } = renderComponent()

    // Search for JPY currency in search bar.
    const searchBar = screen.getByRole("textbox")
    await user.click(searchBar)
    await user.keyboard("JPY")

    // Ensure only JPY button shows and click to change currency.
    const jpyButton = screen.getByRole("button")
    await user.click(jpyButton)

    // Ensure snapshot matches user augmented component.
    expect(asFragment()).toMatchSnapshot()
  })
})
