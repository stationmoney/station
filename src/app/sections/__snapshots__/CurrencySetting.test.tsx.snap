// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Currency Setting loads matches snapshot 1`] = `
<div
  className="grid"
  style={
    Object {
      "gap": 8,
      "gridTemplateColumns": undefined,
      "gridTemplateRows": undefined,
    }
  }
>
  <div
    className="wrapper search search__small"
    style={
      Object {
        "margin": 0,
      }
    }
  >
    <input
      autoComplete="off"
      autoFocus={true}
      className="input input__small"
      inputMode="search"
      onChange={[Function]}
      small={true}
      value=""
    />
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium icon css-i4bv87-MuiSvgIcon-root"
      data-testid="SearchIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
      />
    </svg>
  </div>
  <div
    className="wrapper"
  >
    <section
      className="selector"
    >
      <div
        className="accordion"
      >
        <button
          className="item"
          onClick={[Function]}
        >
          <div
            className="icons_container"
          >
            <div>
              ¥ - Japanese Yen
            </div>
          </div>
          <div
            className="flex track"
            style={
              Object {
                "gap": undefined,
              }
            }
          >
            <span
              className="indicator"
            />
          </div>
        </button>
        <div
          className="content"
        />
      </div>
      <div
        className="accordion"
      >
        <button
          className="item"
          onClick={[Function]}
        >
          <div
            className="icons_container"
          >
            <div>
              ₦ - Nigerian Naira
            </div>
          </div>
          <div
            className="flex track"
            style={
              Object {
                "gap": undefined,
              }
            }
          >
            <span
              className="indicator"
            />
          </div>
        </button>
        <div
          className="content"
        />
      </div>
      <div
        className="accordion"
      >
        <button
          className="item"
          onClick={[Function]}
        >
          <div
            className="icons_container"
          >
            <div>
              $ - United States Dollar
            </div>
          </div>
          <div
            className="flex track"
            style={
              Object {
                "gap": undefined,
              }
            }
          >
            <span
              className="indicator__checked"
            />
          </div>
        </button>
        <div
          className="content"
        />
      </div>
    </section>
  </div>
</div>
`;

exports[`User can change currencies allows user to change currencies 1`] = `
<DocumentFragment>
  <div
    class="grid"
    style="gap: 8px;"
  >
    <div
      class="wrapper search search__small"
      style="margin: 0px;"
    >
      <input
        autocomplete="off"
        class="input input__small"
        inputmode="search"
        value="JPY"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium icon css-i4bv87-MuiSvgIcon-root"
        data-testid="SearchIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
        />
      </svg>
    </div>
    <div
      class="wrapper"
    >
      <section
        class="selector"
      >
        <div
          class="accordion"
        >
          <button
            class="item"
          >
            <div
              class="icons_container"
            >
              <div>
                ¥ - Japanese Yen
              </div>
            </div>
            <div
              class="flex track"
            >
              <span
                class="indicator__checked"
              />
            </div>
          </button>
          <div
            class="content"
          />
        </div>
      </section>
    </div>
  </div>
</DocumentFragment>
`;
